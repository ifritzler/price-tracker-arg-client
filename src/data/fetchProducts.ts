export const fetchProducts = async (
    page: number,
    increased: boolean,
    discountValue: boolean,
    searchValue: string
  ) => {
    return fetch(
      `${process.env.NEXT_PUBLIC_BACKEND_HOST}/api/products?${
        page && `page=${page}`
      }${discountValue ? `&p=true` : ""}${increased ? `&inc=true` : ""}${
        searchValue ? `&q=${searchValue.trim()}` : ""
      }`
    ).then((response) => response.json());
  };