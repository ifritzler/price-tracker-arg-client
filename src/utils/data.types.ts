export type Product = {
  id: number;
  title: string;
  url: string;
  imageUrl: string;
  available: boolean;
  categoryId?: number;
  supermarketId?: number;
  supermarket: {
    id: number;
    name: string;
  };
  category: {
    id: number;
    name: string;
  };
  dailyPrices?: {
    id: number;
    productId: number;
    hasDiscount: boolean;
    price: number;
    discountPrice: number;
    date: Date;
    diffPercentage: number;
  };
};
