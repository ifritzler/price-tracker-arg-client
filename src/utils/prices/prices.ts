export function formatPrice(priceNumber: number): string {
  const priceFormatter = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
  });
  return priceFormatter.format(priceNumber);
}

export enum PriceTypeEnum {
  DISCOUNT,
  STRIKE,
}
