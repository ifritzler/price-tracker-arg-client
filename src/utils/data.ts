import { Product } from "./data.types";

export const testProduct: Product = {
    id: 1,
    title: "Aceite de girasol Cocinero ecobotella 900 cc.",
    url: "https://www.carrefour.com.ar/aceite-de-girasol-cocinero-ecobotella-900-cc-732610/p",
    imageUrl: "https://carrefourar.vtexassets.com/arquivos/ids/422953-1200-auto?v=638373948257230000&width=1200&height=auto&aspect=true",
    available: true,
    categoryId: 1,
    supermarketId: 1,
    supermarket: {
      id: 1,
      name: "carrefour",
    },
    category: {
      id: 1,
      name: "Almac√©n",
    },
    dailyPrices: 
      {
        id: 1,
        productId: 1,
        hasDiscount: true,
        price: 10.99,
        discountPrice: 8.99,
        date: new Date(),
        diffPercentage: 18,
      },
  };