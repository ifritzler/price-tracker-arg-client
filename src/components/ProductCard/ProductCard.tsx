import { Product } from "@/utils/data.types";
import { PriceTypeEnum } from "@/utils/prices/prices";
import Image from "next/image";
import Price from "../Price/Price";
import Button from "../Button/Button";
import { HiOutlineExternalLink } from "react-icons/hi";

interface Props {
  product: Product;
}

export function ProductCard(props: Props) {
  const { product } = props;

  return (
    <article
      className="max-w-[--max-card-with] dark:text-white px-4 py-3 bg-[--background-card] rounded-2xl shadow-md transition-transform hover:scale-105 flex flex-col gap-2 justify-between"
      aria-label="Tarjeta de producto"
    >
      <section
        aria-label="Cabecera de tarjeta de producto"
        className="relative flex flex-col gap-1"
      >
        <Image
          alt={`Imagen del producto: ${product.title}`}
          src={product.imageUrl}
          className="rounded-2xl"
          width={300}
          height={300}
        />
        {product && product.dailyPrices!.diffPercentage > 0 && (
          <span
            aria-label="Faja informativa de producto"
            className="absolute flex bottom-20 left-0 w-full h-6 bg-rose-800 z-50 opacity-95 text-xs justify-center items-center unselectable"
          >
            Aumento del {Number(product.dailyPrices!.diffPercentage).toFixed(0)}
            % en la última jornada
          </span>
        )}
        <Button
          text=""
          typeb="button-link"
          url={`${product.url}`}
          icon={<HiOutlineExternalLink size={20} />}
          onlyIcon
          className="absolute w-fit rounded-md top-3 right-3"
          aria-label="Ir a la pagina web del producto"
        />
      </section>
      <section
        aria-label="Cuerpo con información de la tarjeta de producto"
        className="flex flex-col gap-1 h-full justify-between"
      >
        <h2 aria-label="Título de producto" className="font-medium">
          {product.title}
        </h2>
        <div>
          <div className="flex gap-4 text-xs justify-between">
            <span className="uppercase">{product.supermarket.name}</span>
            <span className="uppercase">{product.category.name}</span>
          </div>
          <div className="flex gap-4">
            {/* Daily prices should only appear if the dailyPrices prop exist */}
            {product.dailyPrices && (
              <section className="flex gap-4 items-center justify-between w-full">
                <span className="font-medium">Precio:</span>
                <Price
                  price={product.dailyPrices.price}
                  type={
                    product.dailyPrices.hasDiscount
                      ? PriceTypeEnum.STRIKE
                      : undefined
                  }
                />
                {product.dailyPrices.hasDiscount && (
                  <Price
                    price={product.dailyPrices.discountPrice}
                    type={PriceTypeEnum.DISCOUNT}
                  />
                )}
              </section>
            )}
          </div>
        </div>
      </section>
      <section
        aria-label="Pie de tarjeta de producto"
        className="flex flex-col gap-1"
      >
        <Button
          text="Ver estadísticas"
          typeb="button-link"
          url={`/product/${product.id}`}
        />
      </section>
    </article>
  );
}
