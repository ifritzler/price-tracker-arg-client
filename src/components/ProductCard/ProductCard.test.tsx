import { testProduct } from "@/utils/data";
import { Product } from "@/utils/data.types";
import { formatPrice } from "@/utils/prices/prices";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { ProductCard } from "./ProductCard";

describe("Product Card", () => {
  it("renders product card unchanged", () => {
    const { container } = render(<ProductCard product={testProduct} />);
    expect(container).toMatchSnapshot();
  });

  it("should be render", () => {
    render(<ProductCard product={testProduct} />);

    const card = screen.getByRole("article");

    expect(card).toBeInTheDocument();
  });

  it("props are well defined", () => {
    render(<ProductCard product={testProduct} />);

    const image = screen.getByAltText(
      `Imagen del producto: ${testProduct.title}`
    );
    const title = screen.getByLabelText("Título de producto");

    expect(image).toBeInTheDocument();
    expect(title).toHaveTextContent(testProduct.title);
  });

  it("price with discount shows correctly", () => {
    render(<ProductCard product={testProduct} />);

    const price = screen.getByLabelText("Precio anterior");
    const discountPrice = screen.getByLabelText(
      "Precio de promoción del producto"
    );

    expect(price).toHaveTextContent(
      formatPrice(testProduct.dailyPrices![0].price)
    );
    expect(discountPrice).toHaveTextContent(
      formatPrice(testProduct.dailyPrices![0].promoPrice)
    );
  });

  it("should only shows price if hasPromotion equals false", () => {
    const mockProduct: Product = {
      ...testProduct,
      dailyPrices: [{ ...testProduct.dailyPrices![0], hasPromotion: false }],
    };
    render(<ProductCard product={mockProduct} />);

    const price = screen.getByLabelText("Precio de lista del producto");

    expect(price).toBeInTheDocument();

    const prices = screen.getAllByTestId("Precio");
    expect(prices.length).toBe(1);
  });

  it("should see both price and discount price if hasPromotion is true", () => {
    render(<ProductCard product={testProduct} />);

    const price = screen.getByLabelText("Precio anterior");
    const discount = screen.getByLabelText("Precio de promoción del producto");

    const prices = screen.getAllByTestId("Precio");
    expect(prices.length).toBe(2);

    expect(price).toBeInTheDocument();
    expect(discount).toBeInTheDocument();
  });

  it("styles shows correctly in case of discount", () => {
    render(<ProductCard product={testProduct} />);

    const price = screen.getByLabelText("Precio anterior");
    const discount = screen.getByLabelText("Precio de promoción del producto");

    expect(price.classList).toContain("text-[--color-text-disabled]");
    expect(discount.classList).toContain("text-[--color-discount]");
  });

  it("should have 2 links", () => {
    render(<ProductCard product={testProduct} />);

    const button = screen.getAllByRole("link");

    expect(button.length).toBe(2);
  });

  it("should have a flag with information within when price has incresed", () => {
    render(<ProductCard product={testProduct} />);

    const flag = screen.getByLabelText("Faja informativa de producto");

    expect(flag).toBeInTheDocument();
  });

  it("flag must render the correct porcentage of increse", () => {
    render(<ProductCard product={testProduct} />);

    const flag = screen.getByLabelText("Faja informativa de producto");

    expect(flag).toHaveTextContent("Aumento del 18% en la última jornada");
  });

  it("flag shouldn't be appear if the percentage of increse is 0", () => {
    render(
      <ProductCard
        product={{
          ...testProduct,
          dailyPrices: [
            {
              ...testProduct.dailyPrices![0],
              diffPercentage: 0, // this is the object of the test
            },
          ],
        }}
      />
    );

    screen.queryByLabelText('Faja informativa de producto')
    const flag = screen.queryByLabelText('Faja informativa de producto')

    expect(flag).toBe(null);
  });
});
