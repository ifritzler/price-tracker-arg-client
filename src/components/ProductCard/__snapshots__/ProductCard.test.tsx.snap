// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card renders product card unchanged 1`] = `
<div>
  <article
    aria-label="Tarjeta de producto"
    class="max-w-[--max-card-with] dark:text-white px-4 py-3 bg-[--background-card] rounded-2xl shadow-md transition-transform hover:scale-105 flex flex-col gap-2 justify-between"
  >
    <section
      aria-label="Cabecera de tarjeta de producto"
      class="relative flex flex-col gap-1"
    >
      <img
        alt="Imagen del producto: Aceite de girasol Cocinero ecobotella 900 cc."
        class="rounded-2xl"
        data-nimg="1"
        decoding="async"
        height="300"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fcarrefourar.vtexassets.com%2Farquivos%2Fids%2F422953-1200-auto%3Fv%3D638373948257230000%26width%3D1200%26height%3Dauto%26aspect%3Dtrue&w=640&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fcarrefourar.vtexassets.com%2Farquivos%2Fids%2F422953-1200-auto%3Fv%3D638373948257230000%26width%3D1200%26height%3Dauto%26aspect%3Dtrue&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fcarrefourar.vtexassets.com%2Farquivos%2Fids%2F422953-1200-auto%3Fv%3D638373948257230000%26width%3D1200%26height%3Dauto%26aspect%3Dtrue&w=640&q=75 2x"
        style="color: transparent;"
        width="300"
      />
      <span
        aria-label="Faja informativa de producto"
        class="absolute flex bottom-20 left-0 w-full h-6 bg-rose-800 z-10 opacity-95 text-xs justify-center items-center unselectable"
      >
        Aumento del 
        18
        % en la última jornada
      </span>
      <a
        aria-label="Ir a la pagina web del producto"
        class="rounded-md p-2 flex items-center justify-center bg-[--color-primary] shadow-sm text-white absolute w-fit rounded-md top-3 right-3 false flex flex-row gap-1"
        href="https://www.carrefour.com.ar/aceite-de-girasol-cocinero-ecobotella-900-cc-732610/p"
        target="_blank"
      >
        <i>
          <svg
            aria-hidden="true"
            fill="none"
            height="20"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 6H6a2 2 0 00-2 2v10a2 2 0 002 2h10a2 2 0 002-2v-4M14 4h6m0 0v6m0-6L10 14"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </i>
      </a>
    </section>
    <section
      aria-label="Cuerpo con información de la tarjeta de producto"
      class="flex flex-col gap-1 h-full justify-between"
    >
      <h2
        aria-label="Título de producto"
        class="font-medium"
      >
        Aceite de girasol Cocinero ecobotella 900 cc.
      </h2>
      <div>
        <div
          class="flex gap-4 text-xs justify-between"
        >
          <span
            class="uppercase"
          >
            carrefour
          </span>
          <span
            class="uppercase"
          >
            Almacén
          </span>
        </div>
        <div
          class="flex gap-4"
        >
          <section
            class="flex gap-4 items-center justify-between w-full"
          >
            <span
              class="font-medium"
            >
              Precio:
            </span>
            <span
              aria-label="Precio anterior"
              class="font-medium text-sm line-through text-[--color-text-disabled]"
              data-testid="Precio"
            >
              $10.99
            </span>
            <span
              aria-label="Precio de promoción del producto"
              class="font-bold text-sm text-[--color-discount]"
              data-testid="Precio"
            >
              $8.99
            </span>
          </section>
        </div>
      </div>
    </section>
    <section
      aria-label="Pie de tarjeta de producto"
      class="flex flex-col gap-1"
    >
      <a
        class="rounded-md px-4 py-2 flex items-center justify-center bg-[--color-primary] shadow-sm text-white undefined w-full flex flex-row gap-1"
        href="/product/1"
        target="_blank"
      >
        <span>
          Ver estadísticas
        </span>
      </a>
    </section>
  </article>
</div>
`;
